include/realtime_tools/realtime_box.h

```diff
@@ -26,50 +26,402 @@
 // ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 // POSSIBILITY OF SUCH DAMAGE.
 
-// Author: Stuart Glaser
+// Author: Lennart Nachtigall
 
-#ifndef REALTIME_TOOLS__REALTIME_BOX_H__
-#define REALTIME_TOOLS__REALTIME_BOX_H__
+#ifndef REALTIME_TOOLS__REALTIME_BOX_H_
+#define REALTIME_TOOLS__REALTIME_BOX_H_
 
+#include <initializer_list>
 #include <mutex>
-#include <string>
+#include <optional>
+#include <utility>
+
+#include <rcpputils/pointer_traits.hpp>
 
 namespace realtime_tools
 {
+
+template <typename T>
+constexpr auto is_ptr_or_smart_ptr = rcpputils::is_pointer<T>::value;
+
 /*!
+    A Box that ensures thread safe access to the boxed contents.
+    Access is best effort. If it can not lock it will return.
+
+    NOTE about pointers:
+    You can use pointers with this box but the access will be different.
+    Only use the get/set methods that take function pointer for accessing the internal value.
+*/
+
+// Provide a base template for the class
+template <class T, typename mutex_type = std::mutex, bool = is_ptr_or_smart_ptr<T>>
+class RealtimeBox;
+
+// Provide a specialisation for non pointer types
+// NOTE: When migrating to a safe access only version just remove the specialisation for pointer
+// and let this be the only version!
+template <class T, typename mutex_type>
+class RealtimeBox<T, mutex_type, false>
+{
+  static_assert(!is_ptr_or_smart_ptr<T>);
+  static_assert(
+    std::is_same_v<mutex_type, std::mutex> || std::is_same_v<mutex_type, std::recursive_mutex>);
+  static_assert(std::is_copy_constructible_v<T>, "Passed type must be copy constructible");
+
+public:
+  using mutex_t = mutex_type;
+  using type = T;
+  // Provide various constructors
+  constexpr explicit RealtimeBox(const T & init = T{}) : value_(init) {}
+  constexpr explicit RealtimeBox(const T && init) : value_(std::move(init)) {}
+
+  // Only enabled for types that can be constructed from an initializer list
+  template <typename U = T>
+  constexpr RealtimeBox(
+    const std::initializer_list<U> & init,
+    std::enable_if_t<std::is_constructible_v<U, std::initializer_list>>)
+  : value_(init)
+  {
+  }
+
+  /**
+   * @brief set a new content with best effort
+   * @return false if mutex could not be locked
+   * @note disabled for pointer types
+   */
+  template <typename U = T>
+  typename std::enable_if_t<!is_ptr_or_smart_ptr<U>, bool> trySet(const T & value)
+  {
+    std::unique_lock<mutex_t> guard(lock_, std::defer_lock);
+    if (!guard.try_lock()) {
+      return false;
+    }
+    value_ = value;
+    return true;
+  }
+  /**
+   * @brief access the content readable with best effort
+   * @return false if the mutex could not be locked
+   * @note only safe way to access pointer type content (rw)
+   */
+  bool trySet(const std::function<void(T &)> & func)
```

Comment:
```suggestion
  bool try_set(const std::function<void(T &)> & func)
```
we've been following snake-case in ros2_control, please adjust the rest too

Reply:
To be honest I am a bit puzzled about this comment given that the PR that introduced these names was accepted https://github.com/ros-controls/realtime_tools/pull/139

This PR ist just for replacing the old RealtimeBox with the new implementation while still keeping it backward compatible with the corresponding deprecation warnings.